### PR TITLE
[dynamo][guards-cpp-refactor] TENSOR_MATCH guard

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120469
* #120416
* #120359
* #120344
* __->__ #120342
* #120096
* #120093
* #120123
* #120119
* #120091
* #120089
* #120068
* #120067
* #120065
* #120064
* #120062
* #120061
* #120060
* #119833
* #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -1006,6 +1006,82 @@ class DATA_PTR_MATCH : public LeafGuard {
...
+    _tensor_check = std::make_unique<TensorCheck>(
+        state,
+        Py_TYPE(item),
+        std::move(tensor),
+        std::move(tensor_dims_size),
+        std::move(tensor_dims_stride));
+  }
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    LocalState state;
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:1047]: I believe constructing LocalState does a bunch of work, including querying TLS.  We should stash this on the root_manager and only construct it once per check.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp]: Nice! Just made the change.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**